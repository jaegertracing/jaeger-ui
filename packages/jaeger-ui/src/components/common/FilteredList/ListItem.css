/*
Copyright (c) 2019 Uber Technologies, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.FilteredList--ListItem {
  align-items: center;
  cursor: pointer;
  display: flex;
  padding: 0.5em 5em 0.5em 0.5em;
  white-space: nowrap;
  min-width: 100%;
}

.FilteredList--ListItem.is-striped {
  background: linear-gradient(to right, rgba(0, 0, 0, 0.04), rgba(0, 0, 0, 0));
}

.FilteredList--ListItem.is-striped:hover,
.FilteredList--ListItem:hover {
  background: linear-gradient(
    to right,
    rgba(135, 232, 222, 0.1),
    rgba(135, 232, 222, 0.2) 70%,
    rgba(135, 232, 222, 0)
  );
  color: var(--tx-color-title);
}

.FilteredList--ListItem.is-selected {
  background: linear-gradient(to right, #87e8de, #87e8deaa 70%, #87e8de00);
  color: var(--tx-color-title);
}

.FilteredList--ListItem.is-striped.is-focused,
.FilteredList--ListItem.is-focused {
  background: linear-gradient(
    to right,
    rgba(135, 232, 222, 0.2),
    rgba(17, 153, 153, 0.2) 70%,
    rgba(17, 153, 153, 0)
  );
  color: var(--tx-color-title);
}

.FilteredList--ListItem mark {
  background: none;
  color: #eb2f96;
  font-weight: 500;
  padding: 0;
}
