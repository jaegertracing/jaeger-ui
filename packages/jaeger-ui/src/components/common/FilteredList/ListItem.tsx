// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import cx from 'classnames';
import { ListChildComponentProps } from 'react-window';

import highlightMatches from './highlightMatches';

import './ListItem.css';

interface IListItemProps extends ListChildComponentProps {
  data: {
    focusedIndex: number | null;
    highlightQuery: string;
    options: string[];
    selectedValue: string | null;
    setValue: (value: string) => void;
  };
}

export default class ListItem extends React.PureComponent<IListItemProps> {
  onClicked = () => {
    const { data, index } = this.props;
    const { options, setValue } = data;
    const value = options[index];
    setValue(value);
  };

  render() {
    const { data, index, style: styleOrig } = this.props;
    // omit the width from the style so the panel can scroll horizontally
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const { width: _, ...style } = styleOrig;
    const { focusedIndex, highlightQuery, options, selectedValue } = data;
    const cls = cx('FilteredList--ListItem', {
      'is-focused': index === focusedIndex,
      'is-selected': options[index] === selectedValue,
      'is-striped': index % 2,
    });
    return (
      <div
        className={cls}
        style={style}
        onClick={this.onClicked}
        role="switch"
        aria-checked={index === focusedIndex ? 'true' : 'false'}
      >
        {highlightMatches(highlightQuery, options[index])}
      </div>
    );
  }
}
