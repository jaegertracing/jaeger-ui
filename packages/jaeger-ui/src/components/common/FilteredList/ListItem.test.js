// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import React from 'react';
import { shallow } from 'enzyme';

import ListItem from './ListItem';

describe('<ListItem>', () => {
  let wrapper;
  let props;
  let setValue;

  beforeEach(() => {
    setValue = jest.fn();
    props = {
      style: {},
      index: 0,
      data: {
        setValue,
        focusedIndex: null,
        highlightQuery: '',
        options: ['a', 'b'],
        selectedValue: null,
      },
    };
    wrapper = shallow(<ListItem {...props} />);
  });

  it('renders without exploding', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('is focused when the index is the focusIndex', () => {
    const data = { ...props.data, focusedIndex: props.index };
    wrapper.setProps({ data });
    expect(wrapper).toMatchSnapshot();
  });

  it('is selected when options[index] == selectedValue', () => {
    const data = { ...props.data, selectedValue: props.data.options[props.index] };
    wrapper.setProps({ data });
    expect(wrapper).toMatchSnapshot();
  });

  it('sets the value when clicked', () => {
    expect(setValue.mock.calls.length).toBe(0);
    wrapper.simulate('click');
    expect(setValue.mock.calls).toEqual([[props.data.options[props.index]]]);
  });
});
