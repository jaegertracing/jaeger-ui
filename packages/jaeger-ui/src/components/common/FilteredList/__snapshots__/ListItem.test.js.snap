// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ListItem> is focused when the index is the focusIndex 1`] = `
<div
  aria-checked="true"
  className="FilteredList--ListItem is-focused"
  onClick={[Function]}
  role="switch"
  style={Object {}}
>
  a
</div>
`;

exports[`<ListItem> is selected when options[index] == selectedValue 1`] = `
<div
  aria-checked="false"
  className="FilteredList--ListItem is-selected"
  onClick={[Function]}
  role="switch"
  style={Object {}}
>
  a
</div>
`;

exports[`<ListItem> renders without exploding 1`] = `
<div
  aria-checked="false"
  className="FilteredList--ListItem"
  onClick={[Function]}
  role="switch"
  style={Object {}}
>
  a
</div>
`;
